saleor/graphql/core/tests/test_query_cost_validation.py

```diff
@@ -0,0 +1,133 @@
+import graphene
+from django.test import override_settings
+
+
+@override_settings(GRAPHQL_QUERY_MAX_COMPLEXITY=1)
+def test_query_exceeding_cost_limit_fails_validation(
+    api_client,
+    variant_with_many_stocks,
+    channel_USD,
+):
+    query_fields = "\n".join(
+        [
+            "p%s:  productVariant(id: $id, channel: $channel) { id }" % i
+            for i in range(20)
+        ]
+    )
+    query = (
+        """
+        query variantAvailability($id: ID!, $channel: String) {
+            %s
+        }
+    """
+        % query_fields
+    )
+
+    variables = {
+        "id": graphene.Node.to_global_id("ProductVariant", variant_with_many_stocks.pk),
+        "channel": channel_USD.slug,
+    }
+
+    response = api_client.post_graphql(query, variables)
+    json_response = response.json()
+    assert "data" not in json_response
+    query_cost = json_response["extensions"]["cost"]["requestedQueryCost"]
+    assert query_cost > 1
+    assert len(json_response["errors"]) == 1
+    assert json_response["errors"][0]["message"] == (
+        f"The query exceeds the maximum cost of 1. Actual cost is {query_cost}"
+    )
```

Comment:
Are we able to also return some error code here? For example, Github API returns a "type" field for this kind of error; when exceeding their max limit you would get:

```
{
  "errors": [
    {
      "type": "MAX_NODE_LIMIT_EXCEEDED",
      "locations": [
        {
          "line": 19,
          "column": 21
        }
      ],
      "message": "By the time this query traverses to the issues connection, it is requesting up to 1,000,000 possible nodes which exceeds the maximum limit of 500,000."
    }
  ]
}
```
I'm wondering if we can achieve something similar.

Reply:
Not sure. Error from cost validator is reported via `ValidationContext.report_error` which mangles `GraphQLError` instances. I'll give it a second look later today after I'm back from running errands.

Reply:
Discussed this on slack and there will be dedicated error code `QueryCostError` in exceptions.